### ISSUE TITLE
Add missing uses of x-ray climb

### ISSUE THREAD

**Issue by fdtaillefer** [miketrethewey/sm-json-data#164]
So far, use of the x-ray climb tech hasn't been elaborated at all. It would be nice to have it at some point.

What's more, x-ray climb is setup from the adjacent room. It can be done when standing in the doorway, so no `adjacentRunway` is needed, but the ability to go to through the door and back is needed. We can represent that with a `resetRoomAtNode`.

**Comment by osse101** [comment]
There are a couple different methods of doing the setup part of the climb.  stationary spin jump, enemy damage boost while going through a door, and shinesparking the frame you transition rooms.  I think we should just add in the spin jump method since it works everywhere (where the geography fits) while the others are very situational.
left wall climbs:  spin jump works everywhere except landing site
right wall climbs: spin jump method takes 3 frame perfect inputs, otherwise needs a dboost or spark.

I put a list together of rta viable spin jump xray climbs (all left wall climbs)
[SuperMetroidXrayClimables.txt](https://github.com/miketrethewey/sm-json-data/files/2923255/SuperMetroidXrayClimables.txt)

**Comment by fdtaillefer** [comment]
Here is what nodes those X-Ray climbs link together, omitting the Crateria one that causes a softlock and the Crab Shaft one that turned out to be impossible:

- Etocoon Energy Tank Room, node 3 to 6
- Green Brinstar Main Shaft, node 13 to 12 with a reset at 8
- Big Pink, node 5 to 4. This one has a problem because node 5 has a `spawnAt` that would take the x-ray climb effectively out of logic.
- Contruction zone, node 3 to 4
- Red Tower, node 2 to 1
- Sporespawn Super Room, node 2 to 1
- Crateria Super Room, node 2 to 1
- Parlor and Alcatraz, node 2 to 8
- Screw Attack Room, node 1 to 2
- Screw Attack Room, node 1 to 3
- Fast Pillars Setup Room, node it 5 to 1 with a reset at 2
- Wasteland, node 1 to 4)
- Worst Room in the Game, node 2 to 4
- Oasis, part of the reqs to go from node 5 to 4. Has a reset at 1.
- Crab Hole, node 2 to 1
- Pants Room, node 5 to 2 with a reset at 1. 1 to 5 has requirements, but they should be applied to the x-ray climb as well so it works out.
- Aqueduct, node 2 to 1
- Aquaduct, node 1 to 9
- Halfie Climb Room, node 2 to 1
- West Cactus Alley Room, node 1 to 2
- Precious Room node 2 to 1 (already in, but lacks the reset at node 2)
- Ice Beam Gate Room, node 2 to 1
- Bubble Mountain, node 5 to 3 with a reset at 4
- Bubble Mountain, node 9 to 1 with a reset at 2. Guess a reset at 3 would also work

**Comment by osse101** [comment]
update to the list:

- Sporespawn Super Room and Criteria Super Room were removed for being out of bounds.
- Fast Pillars Setup Room removed because there is a slope over the door making it an advanced xray climb

For big pink, I feel falling through the crumble blocks is likely enough that we would have to first confirm there is a loop back to the other side of the node 5 door.

**Comment by fdtaillefer** [comment]
The problem with Big Pink is that currently, node 5 has `"spawnAt": 13` which is a flag that indicates that Samus should be considered to spawn at node 13 when entering through 5. Since there's no path back to 5 through the room, currently that means the model offers no way for Samus to ever "be" at 5 to perform the x-ray climb.

Looping back around would still "spawn" her back at 13. So, we will have to figure out a way to make those settings coexist.

**Comment by osse101** [comment]
I figure there should be 4 kinds of xray climb.

1. canXRayClimb - stationary spin jump setup
2. canAdvancedXRayClimb - dboost / shinespark / triple frame perfect door grab
3. canOOBXRayClimb - leave the map boundary then come back in to hit an inbounds door
4. canOOBWithXRay - leave the map boundary to interact with OOB tiles/data

**Comment by fdtaillefer** [comment]
At a glance, that sounds like good separation. Here's my view on each:
1. Should definitely be added to the project.
2. and 3. I don't think we *need* them anytime soon, but I don't think it would hurt to have them if someone wants to do them.
4. A scary one. I don't know a lot about major glitches, but I'm worried our model might break down and be unable to properly represent all the convoluted links this one may need. I'd advise caution and discussion before starting work on these.

**Comment by fdtaillefer** [comment]
As an aside, I think the Big Pink issue could be solved if we added one more node between 5 and 13. We can make Samus spawn there, but the X-Ray climb on that one, but barring the X-Ray climb it can only lead you to node 13.

The new node wouldn't strictly match an actual spot in the room, but it would work logically.

**Comment by osse101** [comment]
That's exactly how I feel about the breakdown of X-Ray climbs.

The new node would then be there purely to counteract the `spawnAt`.  If we don't want to remove the `spawnAt`, then it sounds reasonable but would be very similar to what you'd expect from the door node itself.

**Comment by fdtaillefer** [comment]
It is: the main point of the `spawnAt` is to prevent the logic from expecting you to be able to go back through door 5 under normal circumstances.

Because you can't get to that spot except by exiting through door 5, I think we don't need to specify a `resetAtnode` for this one.

One issue though is this solution does nothing to require the player to make a loop around to x-ray climb the second time around. Don't know how big of a problem that is though.

**Comment by fdtaillefer** [comment]
Lots of work done on this by #174.

Maybe we could close this and create separate issues for Big Pink and for other x-ray climb techs we may want?
